teps/0111-replace-volumes-with-workspaces.md

```diff
@@ -0,0 +1,368 @@
+
+Cons:
+- There's no clear use case for many of these types of Volumes in Tekton.
+
+#### Support hostPath volumes in Workspace Bindings
+
+This solution augments the proposed solution (removing volume-related fields from the API)
+with new Workspace Bindings to support advanced Docker-in-Docker use cases.
+
+In this solution, a new field be added to [WorkspaceBinding](https://github.com/tektoncd/pipeline/blob/342ed5237f3bd3273485672426472194a186c02f/pkg/apis/pipeline/v1beta1/workspace_types.go#L54)
+to support hostPath volumes:
+
+```go
+import corev1 "k8s.io/api/core/v1"
+
+type WorkspaceBinding struct {
+	// Name is the name of the workspace populated by the volume.
+	Name string `json:"name"`
+	// SubPath is an optional directory on the volume which should be used for this binding
+	SubPath string `json:"subPath,omitempty"`
+  
+  // ... existing volume sources such as emptyDir, secret, PVC
+
+  // HostPath represents a pre-existing file or directory on the host
+	// machine that is directly exposed to the container.
+  HostPath *corev1.HostPathVolumeSource
+}
+```
+
+Pros:
+- Supports more advanced docker-in-docker use cases.
+
+Cons:
+- Not necessary for most CI/CD use cases.
+- Tasks that require hostPath would likely want to declare this dependency at authoring time rather than at runtime.
+- Task authors may not want their Tasks to be used with hostPath workspace bindings due to the
+[security risks](https://blog.quarkslab.com/kubernetes-and-hostpath-a-love-hate-relationship.html)
+they pose.
+
+#### Support CSI volumes in Workspace Bindings
+
+Kubernetes has implemented support for volumes conforming to the [Container Storage Interface (CSI)]((https://kubernetes.io/blog/2019/01/15/container-storage-interface-ga/),
+allowing vendors to develop their own storage plugins for use with Kubernetes.
+
+[CSI volumes](https://kubernetes.io/docs/concepts/storage/volumes/#csi) can be used in a Pod via a PersistentVolumeClaim or
+a generic or CSI-specific ephemeral volume.
+Workspace Bindings already support PersistentVolumeClaims, meaning that CSI volumes can be used in Workspaces.
+We may want to support [CSI ephemeral volumes](https://kubernetes.io/docs/concepts/storage/ephemeral-volumes/#csi-ephemeral-volumes)
+as workspace backings as well. One use case for CSI ephemeral volumes is to be able to use Hashicorp Vault to mount secrets
+into a Task's pod, as requested in [#4446](https://github.com/tektoncd/pipeline/issues/4446).
+
+In the future, if users request it, we can also consider supporting
+[generic ephemeral volumes](https://kubernetes.io/docs/concepts/storage/ephemeral-volumes/#generic-ephemeral-volumes).
+
+### Use TaskRun.Spec.PodTemplate to support missing Workspace features provided by Volumes
+
+A [proposed alternative](https://github.com/tektoncd/pipeline/issues/2057#issuecomment-587559783)
+to implementing support for extra volume types as workspace bindings is to use the `podTemplate` field
```

Comment:
If we are going to remove volumemount and volumes for tasks, will there be a document on how to migrate podTemplate.volumes with workspaces?

Reply:
Definitely! I added a section about this under "design considerations".